src/Propel/Runtime/ActiveQuery/Criterion/ExistsCriterion.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/**
+ * MIT License. This file is part of the Propel package.
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Propel\Runtime\ActiveQuery\Criterion;
+
+use Propel\Runtime\ActiveQuery\ModelJoin;
+use Propel\Runtime\Map\RelationMap;
+
+/**
+ * Specialized Criterion used for EXISTS
+ */
+class ExistsCriterion extends AbstractCriterion
+{
+    public const TYPE_EXISTS = 'EXISTS';
+
+    public const TYPE_NOT_EXISTS = 'NOT_EXISTS';
+
+    /**
+     * The inner query of the exists
+     *
+     * @var \Propel\Runtime\ActiveQuery\ModelCriteria
+     */
+    private $existsQuery;
+
+    /**
+     * Build NOT EXISTS instead of EXISTS
+     *
+     * @var string $keyword Either ExistsCriterion::TYPE_EXISTS or ExistsCriterion::TYPE_NOT_EXISTS
+     *
+     * @phpstan-var ExistsCriterion::TYPE_*
+     */
+    private $type = self::TYPE_EXISTS;
+
+    /**
+     * @phpstan-param ExistsCriterion::TYPE_*|null $type
+     *
+     * @param \Propel\Runtime\ActiveQuery\ModelCriteria $outerQuery
+     * @param \Propel\Runtime\ActiveQuery\ModelCriteria $existsQuery
+     * @param string|null $type Either ExistsCriterion::TYPE_EXISTS or ExistsCriterion::TYPE_NOT_EXISTS
+     * @param \Propel\Runtime\Map\RelationMap|null $relationMap where outer query is on the left side
+     */
+    public function __construct($outerQuery, $existsQuery, ?string $type = null, ?RelationMap $relationMap = null)
+    {
+        parent::__construct($outerQuery, '', null, null);
+        $this->existsQuery = $existsQuery;
+        $this->type = ($type !== null && $type === self::TYPE_NOT_EXISTS) ? self::TYPE_NOT_EXISTS : self::TYPE_EXISTS;
+
+        if ($relationMap !== null) {
+            $joinCondition = $this->buildJoinCondition($outerQuery, $relationMap);
+            $this->existsQuery->addAnd($joinCondition);
+        }
+    }
+
+    /**
+     * @see \Propel\Runtime\ActiveQuery\Criterion\AbstractCriterion::appendPsForUniqueClauseTo()
+     *
+     * @param string $sb The string that will receive the Prepared Statement
+     * @param array $params A list to which Prepared Statement parameters will be appended
+     *
+     * @return void
+     */
+    protected function appendPsForUniqueClauseTo(&$sb, array &$params)
+    {
+        $existsQuery = $this->existsQuery
+            ->clearSelectColumns()
+            ->addAsColumn('existsFlag', '1')
+            ->createSelectSql($params);
+        if ($this->type === self::TYPE_NOT_EXISTS) {
+            $sb .= 'NOT ';
+        }
+        $sb .= 'EXISTS (' . $existsQuery . ')';
```

Comment:
This string-building operation can probably be simplified as such: `$sb .= $this->type . ' (' . $existsQuery . ')';`

Reply:
Right, I'll change the TYPE_NOT_EXISTS so that this works

Reply:
Thinking about it, having this depend on the type token being the correct SQL operator is not optimal, but now I have changed it already and it probably won't ever matter